Ruby Deprecation doc (#2813)
Azure/azure-sdk-for-ruby: docs/README.md

# Microsoft Azure SDK for Ruby

> Microsoft Azure SDK for Ruby has been deprecated. It will be in maintenance mode for a period of time. This doc instructs how to use Azure REST API as a replacement. By the way, it is highly recommanded you to use GA SDKs

Reviewer: ```suggestion
As of (2022/01), Azure SDK for Ruby - Resource Management Libraries are officially deprecated. The libraries will be in maintenance mode for a period of 3 years and security / bug fixes will still be applied. However, there will be no further releases based on feature requests. 

We recommend you take one of the following approaches
- Use Azure REST API directly in Ruby (instructions are provided in this doc)
- Use one of the next-generation SDKs that provides more idiomatic, consistent, approachable, dependable, and diagnosable experience. You can find more info on the repo of each individual language
  - [Java](https://github.com/Azure/azure-sdk-for-java/#management-new-releases)
  - [.NET](https://github.com/Azure/azure-sdk-for-net/#management-new-releases)
  - [Python](https://github.com/Azure/azure-sdk-for-python/#management-new-releases)
```

* [Python](https://github.com/Azure/azure-sdk-for-python)
* [Java](https://github.com/Azure/azure-sdk-for-java)
* [JavaScript](https://github.com/Azure/azure-sdk-for-js)
* [.Net](https://github.com/Azure/azure-sdk-for-net)

> You can also find other SDKs in [Azure SDK](https://github.com/Azure/azure-sdk).

Reviewer: ```suggestion
You can also find information about the latest SDK releases as well as API references on the [official Azure SDK web-site](https://azure.github.io/azure-sdk/)
```

> # Azure REST API

Reviewer: ```suggestion
# Using Azure REST API in Ruby
```

[Azure REST API Reference](https://docs.microsoft.com/en-us/rest/api/azure/)

> 

Reviewer: ```suggestion
## Reference
[Azure REST API Reference](https://docs.microsoft.com/en-us/rest/api/azure/)
```

## Authentication

> The first step to using the SDK is authentication and permissions. For people unfamilar with Azure this may be one of

Reviewer: so we are still relying on ms_rest_azure to use the REST API ? if this is the case, we probably should mention it at the beginning as well

> The first step to using the SDK is authentication and permissions. For people unfamilar with Azure this may be one of

Reviewer: The first step to calling the REST API is authentication. For people who are unfamiliar with Azure, this maybe one of the challenging concepts. 

To authenticate, we will need to set up a Service Principal. For reference on how to set up a Service Principal from command line, please see [Authenticating a service principal with Azure Resource Manager](http://aka.ms/cli-service-principal) or [Unattended Authentication](http://aka.ms/auth-unattended).

For a more robust explanation of authentication in Azure, please follow  [Developer’s guide to auth with Azure Resource Manager API](http://aka.ms/arm-auth-dev-guide).

After creating the Service Principal, you should have the following key identifiers
- Client ID (GUID)
- Client Secret (String)
- Tenant ID (GUID) or domain name (String)

Reviewer 2: Thanks Nick. It seemed that above link about Developer’s guide to auth with Azure Resource Manager API could not be opened.

the more difficult concepts. For a reference on setting up a service principal from the command line see
[Authenticating a service principal with Azure Resource Manager](http://aka.ms/cli-service-principal) or
[Unattended Authentication](http://aka.ms/auth-unattended). For a more robust explanation of authentication in Azure,
see [Developer’s guide to auth with Azure Resource Manager API](http://aka.ms/arm-auth-dev-guide).

After creating the service principal, you should have three pieces of information, a client id (GUID), client secret
(string) and tenant id (GUID) or domain name (string).

# Usage of ms_rest_azure gem

The ms_rest_azure gem is a Azure rest client library used by Azure SDK for Ruby. It relies on faraday and concurrent-ruby. 

## Install

You can install the ms_rest_azure gem with the following command:

> 

Reviewer: consider moving "The ms_rest_azure gem is a Azure rest client library used by Azure SDK for Ruby. It relies on faraday and concurrent-ruby. " into the Install section

```ruby

> gem install 'ms_rest_azure'

Reviewer: what's the current status of ms_rest_azure, is it also being deprecated?

Reviewer: or is someone maintaining it?

Author: ms_rest_azure is the runtime part of Ruby SDK. I think we can keep maintaining this library. It's stable, light weighted and easy to use.

```

## Prerequisite

In order to use the Azure client, you must supply the following values:

* Tenant Id
* Client Id
* Subscription Id
* Client Secret

You could pass the above values in the following ways:

### Option 1 - Environment Variables
You can set the (above) values using the following environment variables:

* AZURE_TENANT_ID
* AZURE_CLIENT_ID
* AZURE_SUBSCRIPTION_ID
* AZURE_CLIENT_SECRET

To set the environment variables, in Windows, you could use the command such as:

```
set AZURE_TENANT_ID=<YOUR_TENANT_ID>
```

In Unix based systems, you could use the command such as:

```
export AZURE_TENANT_ID=<YOUR_TENANT_ID>
```

### Option 2 - Options Hash

> The initialization of profile clients take an options hash as a parameter. This options hash consists of tenant_id, client_id, client_secret, subscription_id, active_directory_settings and credentials. Among these, the active_directory_settings and credentials are optional.

Reviewer: In addition to environment variables, you can also pass those parameters in an Option hash when you initialize the profile clients

You can set the (above) values using the options hash:

```ruby
options = {
  tenant_id: 'YOUR TENANT ID',
  client_id: 'YOUR CLIENT ID',
  client_secret: 'YOUR CLIENT SECRET',
  subscription_id: 'YOUR SUBSCRIPTION ID'
}
```

## OAuth2 Token Credentials

> [Acquire an acces token](https://docs.microsoft.com/en-us/rest/api/azure/#acquire-an-access-token)

Reviewer: maybe provide a little bit background info on how token works in Azure

You could use the following code

```ruby
provider = MsRestAzure::MSITokenProvider.new()
credentials = MsRest::TokenCredentials.new(provider)
```

AzureCliTokenProvider use [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/) to acquire token. Azure CLI should be installed in PATH. You could use the following code

```ruby
provider = MsRestAzure::AzureCliTokenProvider.new()
credentials = MsRest::TokenCredentials.new(provider)
```

### ActiveDirectoryServiceSettings

ActiveDirectoryServiceSettings class represents a settings for Azure AD authentication. It contains 4 sets of properties for different clouds

* get_azure_settings
* get_azure_china_settings
* get_azure_german_settings
* get_azure_us_government_settings

So you can create a application token provider with Azure China Settings by using the following code

```ruby
settings = MsRestAzure::ActiveDirectoryServiceSettings.get_azure_china_settings
provider = MsRestAzure::ApplicationTokenProvider.new(
       'YOUR TENANT ID',
       'YOUR CLIENT ID',
       'YOUR CLIENT SECRET',
       settings)
credentials = MsRest::TokenCredentials.new(provider)
```

## Make an async request 

> AzureServiceClient can be used to make an async request. You could use it directly or create a class that inherits it.

Reviewer: is async request the standard pattern in Ruby? if so, we should provide an async request code sample

Author: All the samples will be asynchronized.

> ### Example

Reviewer: Code Sample

    base_url: request_url
}

# Make request
promise = client.make_request_async(request_url, :post, path_template, request_options)

promise = promise.then do |result|
  http_response = result.response
  status_code = http_response.status
  response_content = http_response.body

  unless status_code == 200
    error_model = JSON.load(response_content)
    fail MsRestAzure::AzureOperationError.new(result.request, http_response, error_model)
  end

  result.request_id = http_response['x-ms-request-id'] unless http_response['x-ms-request-id'].nil?
  result.correlation_request_id = http_response['x-ms-correlation-request-id'] unless http_response['x-ms-correlation-request-id'].nil?
  result.client_request_id = http_response['x-ms-client-request-id'] unless http_response['x-ms-client-request-id'].nil?
  
  # Deserialize Response
  if status_code == 200
    begin
      result.body = response_content.to_s.empty? ? nil : JSON.load(response_content)
    rescue Exception => e
      fail MsRest::DeserializationError.new('Error occurred in deserializing the response', e.message, e.backtrace, result)
    end
  end

  result
end

response = promise.execute.value!

```

> 

Reviewer: we should provide one more sample in addition to checking storage availability. A classical sample is creating a virtual machine as it contains creating a resource group, and network

Author: I will add additional markdown files for that.

> # Provide feedback

Reviewer: refine this section to be more like https://github.com/Azure/azure-sdk-for-java/#need-help

# Microsoft Azure SDK for Ruby

As of (2022/01), Azure SDK for Ruby - Resource Management Libraries are officially deprecated. The libraries will be in maintenance mode for a period of 3 years and security / bug fixes will still be applied. However, there will be no further releases based on feature requests. 

> We recommend you use Azure REST API directly in Ruby (instructions are provided in this doc)

Reviewer 3: "In the future, we recommend that you interface with the Azure REST APIs directly from Ruby; instructions are provided in this doc."

# Microsoft Azure SDK for Ruby

As of (2022/01), Azure SDK for Ruby - Resource Management Libraries are officially deprecated. The libraries will be in maintenance mode for a period of 3 years and security / bug fixes will still be applied. However, there will be no further releases based on feature requests. 

We recommend you take one of the following approaches
- Use Azure REST API directly in Ruby (instructions are provided in this doc)

> - Use one of the next-generation SDKs that provides more idiomatic, consistent, approachable, dependable, and diagnosable experience. You can find more info on the repo of each individual language

Reviewer 3: No we can't say that. If customers are using Ruby, we can't tell them to switch language

Author: > No we can't say that. If customers are using Ruby, we can't tell them to switch language

OK. I will remove this paragraph.

> We recommend you take one of the following approaches
> - Use Azure REST API directly in Ruby (instructions are provided in this doc)
> - Use one of the next-generation SDKs that provides more idiomatic, consistent, approachable, dependable, and diagnosable experience. You can find more info on the repo of each individual language
>   - [Java](https://github.com/Azure/azure-sdk-for-java/#management-new-releases)
>   - [.NET](https://github.com/Azure/azure-sdk-for-net/#management-new-releases)
>   - [Python](https://github.com/Azure/azure-sdk-for-python/#management-new-releases)

Author: ```suggestion
We recommend you use Azure REST API directly in Ruby (instructions are provided in this doc)
```

* Tenant Id
* Client Id
* Subscription Id
* Client Secret

You could pass the above values in the following ways:

### Option 1 - Environment Variables
You can set the (above) values using the following environment variables:

* AZURE_TENANT_ID
* AZURE_CLIENT_ID
* AZURE_SUBSCRIPTION_ID
* AZURE_CLIENT_SECRET

To set the environment variables, in Windows, you could use the command such as:

```
set AZURE_TENANT_ID=<YOUR_TENANT_ID>
```

In Unix based systems, you could use the command such as:

```
export AZURE_TENANT_ID=<YOUR_TENANT_ID>
```

### Option 2 - Options Hash
In addition to environment variables, you can also pass those parameters in an Option Hash when you initialize the profile clients. This options hash consists of tenant_id, client_id, client_secret, subscription_id, active_directory_settings and credentials. Among these, the active_directory_settings and credentials are optional.

You can set the (above) values using the options hash:

```ruby
options = {
  tenant_id: 'YOUR TENANT ID',
  client_id: 'YOUR CLIENT ID',
  client_secret: 'YOUR CLIENT SECRET',
  subscription_id: 'YOUR SUBSCRIPTION ID'
}
```

## OAuth2 Token Credentials

> ### Step 1 [Acquire an acces token](https://docs.microsoft.com/en-us/rest/api/azure/#acquire-an-access-token)

Reviewer 4: ```suggestion
### Step 1 [Acquire an access token](https://docs.microsoft.com/en-us/rest/api/azure/#acquire-an-access-token)
```

# Microsoft Azure SDK for Ruby

As of (2022/01), Azure SDK for Ruby - Resource Management Libraries are officially deprecated. The libraries will be in maintenance mode for a period of 3 years and security / bug fixes will still be applied. However, there will be no further releases based on feature requests. 

We recommend you take one of the following approaches
- Use Azure REST API directly in Ruby (instructions are provided in this doc)
- Use one of the next-generation SDKs that provides more idiomatic, consistent, approachable, dependable, and diagnosable experience. You can find more info on the repo of each individual language
  - [Java](https://github.com/Azure/azure-sdk-for-java/#management-new-releases)
  - [.NET](https://github.com/Azure/azure-sdk-for-net/#management-new-releases)
  - [Python](https://github.com/Azure/azure-sdk-for-python/#management-new-releases)
* [Python](https://github.com/Azure/azure-sdk-for-python)
* [Java](https://github.com/Azure/azure-sdk-for-java)
* [JavaScript](https://github.com/Azure/azure-sdk-for-js)

> * [.Net](https://github.com/Azure/azure-sdk-for-net)

Reviewer: we can remove those redundant repo links

You can also find information about the latest SDK releases as well as API references on the [official Azure SDK web-site](https://azure.github.io/azure-sdk/)

# Using Azure REST API

> ## Reference

Reviewer: ```suggestion
## API Reference
```

> 

Reviewer: I suggest moving API reference section to between authentication and making the REST call

You can find the references of all available REST API endpoints on this doc location
[Azure REST API Reference](https://docs.microsoft.com/en-us/rest/api/azure/)

## Authentication

The first step to using the SDK is authentication and permissions. For people unfamilar with Azure this may be one of the more difficult concepts. 

To authenticate, we will need to set up a Service Principal. For reference on how to set up a Service Principal from command line, please see
[Authenticating a service principal with Azure Resource Manager](http://aka.ms/cli-service-principal) or
[Unattended Authentication](http://aka.ms/auth-unattended).

For a more robust explanation of authentication in Azure,
please follow [Developer’s guide to auth with Azure Resource Manager API](http://aka.ms/arm-auth-dev-guide).

After creating the Service Principal, you should have the following key identifiers
* Client ID (GUID)
* Client Secret (String)
* Tenant ID (GUID) or domain name (String).

> ## Make the request

Reviewer: ```suggestion
## Make your REST request
```

### Option 1 - Environment Variables
You can set the (above) values using the following environment variables:

* AZURE_TENANT_ID
* AZURE_CLIENT_ID
* AZURE_SUBSCRIPTION_ID
* AZURE_CLIENT_SECRET

To set the environment variables, in Windows, you could use the command such as:

```
set AZURE_TENANT_ID=<YOUR_TENANT_ID>
```

In Unix based systems, you could use the command such as:

```
export AZURE_TENANT_ID=<YOUR_TENANT_ID>
```

### Option 2 - Options Hash
In addition to environment variables, you can also pass those parameters in an Option Hash when you initialize the profile clients. This options hash consists of tenant_id, client_id, client_secret, subscription_id, active_directory_settings and credentials. Among these, the active_directory_settings and credentials are optional.

You can set the (above) values using the options hash:

```ruby
options = {
  tenant_id: 'YOUR TENANT ID',
  client_id: 'YOUR CLIENT ID',
  client_secret: 'YOUR CLIENT SECRET',
  subscription_id: 'YOUR SUBSCRIPTION ID'
}
```

## OAuth2 Token Credentials

[Acquire an acces token](https://docs.microsoft.com/en-us/rest/api/azure/#acquire-an-access-token)

> ### Use token provider

Reviewer: are token providers mandatory in our scenario?  if so, we should make it more clear like Step 1 -> Step 2 -> Step 3

    base_url: request_url
}

# Make request
promise = client.make_request_async(request_url, :post, path_template, request_options)

promise = promise.then do |result|
  http_response = result.response
  status_code = http_response.status
  response_content = http_response.body

  unless status_code == 200
    error_model = JSON.load(response_content)
    fail MsRestAzure::AzureOperationError.new(result.request, http_response, error_model)
  end

  result.request_id = http_response['x-ms-request-id'] unless http_response['x-ms-request-id'].nil?
  result.correlation_request_id = http_response['x-ms-correlation-request-id'] unless http_response['x-ms-correlation-request-id'].nil?
  result.client_request_id = http_response['x-ms-client-request-id'] unless http_response['x-ms-client-request-id'].nil?
  
  # Deserialize Response
  if status_code == 200
    begin
      result.body = response_content.to_s.empty? ? nil : JSON.load(response_content)
    rescue Exception => e
      fail MsRest::DeserializationError.new('Error occurred in deserializing the response', e.message, e.backtrace, result)
    end
  end

  result
end

response = promise.execute.value!

```

> ## Other Examples

Reviewer: ```suggestion
## Additional Code Samples
```
